src/community/assortativity.jl

```diff
@@ -0,0 +1,52 @@
+"""
+    assortativity(g)
+
+Return the [assortativity coefficient](https://en.wikipedia.org/wiki/Assortativity)
+of graph `g`, defined as the Pearson correlation of excess degree between
+the end vertices of all the edges of the graph.
+
+The excess degree is equal to the degree of linked vertices minus one,
+i.e. discounting the edge that links the pair.
+In directed graphs, the paired values are the out-degree of source vertices
+and the in-degree of destination vertices.
+
+# Examples
+```jldoctest
+julia> using LightGraphs
+
+julia> assortativity(star_graph(4))
+-1.0
+```
+"""
+function assortativity(g::AbstractGraph)
+    nue  = ne(g)
+    sjk = sj = sk = sjs = sks = 0
+    for d in edges(g)
+        j = outdegree(g, src(d)) - 1
+        k = indegree(g, dst(d)) - 1
+        sjk += j*k
+        sj  += j
+        sk  += k
+        sjs += j^2
```

Comment:
do we need any sort of overflow protection/detection here?

Reply:
Those values are used in operations that eventually yield floating point numbers. So they might be defined as `Float64` from the outset, and overflow would be avoided. Is that acceptable?

Reply:
`sjs` is an integer and it's accumulating the squares of degrees. A few hubs on a larger graph on a 32 bit system and it seems you might run into some overflow.

Reply:
Oh, I didn't answer your question. I'd just keep them as integers but I might make them explicitly `Int64` to be safe.

Reply:
https://github.com/JuliaGraphs/LightGraphs.jl/blob/3a6f3a0e9570ecc0018c040b69ddb541b33fbe39/src/community/assortativity.jl#L21-L27

Instead of fixing the type to `Int64`, it is promoted to be *at least* `Int64`. I was tempted to use `UInt64`, since negative values should not happen, but I was refrained by the lines #26 and #27, where `1` is subtracted to the results of `outdegree`and `indegree`. Theoretically those values should always be greater or equal to zero, but just in case...